File: src/main/java/com/chaottic/nbt/number/DoubleTag.java
package com.chaottic.nbt.number;

import java.io.DataInput;
import java.io.IOException;

public record DoubleTag(double d) implements NumberTag {
    @Override
    public byte getAsByte() {
        return (byte) d;
    }

    @Override
    public short getAsShort() {
        return (short) d;
    }

    @Override
    public int getAsInt() {
        return (int) d;
    }

    @Override
    public long getAsLong() {
        return (long) d;
    }

    @Override
    public float getAsFloat() {
        return (float) d;
    }

    @Override
    public double getAsDouble() {
        return d;
    }

    public static DoubleTag createDoubleTag(DataInput dataInput) throws IOException {
        return new DoubleTag(dataInput.readDouble());
    }
}

File: src/main/kotlin/com/chaottic/nbt/number/DoubleTag.kt
package com.chaottic.nbt.number

import com.chaottic.nbt.TagType
import java.io.DataInput

@JvmInline
value class DoubleTag(private val d: Double) : NumberTag {

	override fun asByte() = d.toLong().toByte()
	override fun asShort() = d.toLong().toShort()
	override fun asInt() = d.toInt()
	override fun asLong() = d.toLong()
	override fun asFloat() = d.toFloat()
	override fun asDouble() = d

	companion object : TagType<DoubleTag> {

		override fun createTag(dataInput: DataInput) = DoubleTag(dataInput.readDouble())
	}
}
